core/commands/profile.go

```diff
@@ -0,0 +1,234 @@
+		Tagline: "Collect a performance profile for debugging.",
+		ShortDescription: `
+Collects cpu, heap, and goroutine profiles from a running go-ipfs daemon
+into a single zipfile. To aid in debugging, this command also attempts to
+include a copy of the running go-ipfs binary.
+`,
+		LongDescription: `
+Collects cpu, heap, and goroutine profiles from a running go-ipfs daemon
+into a single zipfile. To aid in debugging, this command also attempts to
+include a copy of the running go-ipfs binary.
+
+Privacy Notice:
+
+The output file includes:
+
+- A list of running goroutines.
+- A CPU profile.
+- A heap profile.
+- Your copy of go-ipfs.
+- The output of 'ipfs version --all'.
+
+It does not include:
+
+- Any of your IPFS data or metadata.
+- Your config or private key.
+- Your IP address.
+- The contents of your computer's memory, filesystem, etc.
+
+However, it could reveal:
+
+- Your build path, if you built go-ipfs yourself.
+- If and how a command/feature is being used (inferred from running functions).
+- Memory offsets of various data structures.
+- Any modifications you've made to go-ipfs.
+`,
+	},
+	NoLocal: true,
+	Options: []cmds.Option{
+		cmds.StringOption(outputOptionName, "o", "The path where the output should be stored."),
+		cmds.StringOption(cpuProfileTimeOption, "The amount of time spent profiling CPU usage.").WithDefault("30s"),
+	},
+	Run: func(req *cmds.Request, res cmds.ResponseEmitter, env cmds.Environment) error {
+		cpuProfileTimeStr, _ := req.Options[cpuProfileTimeOption].(string)
+		cpuProfileTime, err := time.ParseDuration(cpuProfileTimeStr)
+		if err != nil {
+			return fmt.Errorf("failed to parse CPU profile duration %q: %w", cpuProfileTimeStr, err)
+		}
+
+		r, w := io.Pipe()
+		go func() {
+			_ = w.CloseWithError(writeProfiles(req.Context, cpuProfileTime, w))
+		}()
+		return res.Emit(r)
+	},
+	PostRun: cmds.PostRunMap{
+		cmds.CLI: func(res cmds.Response, re cmds.ResponseEmitter) error {
+			v, err := res.Next()
+			if err != nil {
+				return err
+			}
+
+			outReader, ok := v.(io.Reader)
+			if !ok {
+				return e.New(e.TypeErr(outReader, v))
+			}
+
+			outPath, _ := res.Request().Options[outputOptionName].(string)
+			if outPath == "" {
+				outPath = "ipfs-profile-" + time.Now().Format(timeFormat) + ".zip"
+			}
+			fi, err := os.Create(outPath)
+			if err != nil {
+				return err
+			}
+			defer fi.Close()
+
+			_, err = io.Copy(fi, outReader)
+			if err != nil {
+				return err
+			}
+			return re.Emit(&profileResult{File: outPath})
+		},
+	},
+	Encoders: cmds.EncoderMap{
+		cmds.Text: cmds.MakeTypedEncoder(func(req *cmds.Request, w io.Writer, out *profileResult) error {
+			fmt.Fprintf(w, "Wrote profiles to: %s\n", out.File)
+			return nil
+		}),
+	},
+}
+
+func writeProfiles(ctx context.Context, cpuProfileTime time.Duration, w io.Writer) error {
+	archive := zip.NewWriter(w)
+
+	// Take some profiles.
+	type profile struct {
+		name  string
+		file  string
+		debug int
+	}
+
+	profiles := []profile{{
```

Comment:
ditching the contention profile because of libp2p noisiness?

Reply:
Yeah, there's no point for now. We'd also need to leave mutex profiling on for a bit to get anything useful which would make this take even longer.